File: modules/native-audio/ios/NativeAudioModule.swift
import ExpoModulesCore

public class NativeAudioModule: Module {
  private lazy var metronomeEngine: MetronomeEngine = createMetronomeEngine()
  private let clickSoundPlayer = ClickSoundPlayer()

  public func definition() -> ModuleDefinition {
    Name("NativeAudioModule")

    Events("onTick")

    Function("initialize") {
      self.clickSoundPlayer.initialize()
    }

    Function("areSoundsReady") { () -> Bool in
      self.clickSoundPlayer.areReady
    }

    Function("start") { (options: [String: Any]) in
      let bpm = (options["bpm"] as? NSNumber)?.doubleValue ?? 120
      let beatsPerMeasure = (options["beatsPerMeasure"] as? NSNumber)?.intValue ?? 4
      let accentPattern = self.readAccentPattern(
        options["accentPattern"],
        beatsPerMeasure: beatsPerMeasure
      )
      let ticksPerBeat = self.readTicksPerBeat(options["subdivision"])
      let playbackMode = options["playbackMode"] as? String
      let timelineEvents: [TimelinePlaybackEvent] =
        playbackMode == "song_timeline"
          ? self.readTimelineEvents(options["timelineEvents"])
          : []

      // Prepare/calibrate happens inside MetronomeEngine.start (preparing phase)
      // before the future anchor and first lookahead publish.
      self.metronomeEngine.start(
        bpm: bpm,
        beatsPerMeasure: beatsPerMeasure,
        accentPattern: accentPattern,
        ticksPerBeat: ticksPerBeat,
        timelineEvents: timelineEvents
      )
    }

    Function("stop") {
      self.metronomeEngine.stop()
      self.clickSoundPlayer.flushScheduled()
    }

    Function("setTempo") { (bpm: Double) in
      self.metronomeEngine.updateTempo(bpm)
    }

    Function("setAccentPattern") { (accentPattern: [Bool]) in
      self.metronomeEngine.updateAccentPattern(accentPattern)
    }

    Function("setSubdivision") { (subdivision: String) in
      self.metronomeEngine.updateSubdivision(self.readTicksPerBeat(subdivision))
    }

    Function("setSubdivisionAccentMode") { (mode: String) in
      self.metronomeEngine.updateSubdivisionAccentMode(self.readSubdivisionAccentMode(mode))
    }

    Function("setSubdivisionAccentEveryNth") { (everyNth: Int) in
      self.metronomeEngine.updateSubdivisionAccentEveryNth(everyNth)
    }

    Function("setSubdivisionAccentPattern") { (pattern: [Bool]) in
      self.metronomeEngine.updateSubdivisionAccentPattern(pattern)
    }

    Function("setNormalClickSound") { (soundId: String) in
      self.clickSoundPlayer.setNormalClickSound(soundId)
    }

    Function("setAccentClickSound") { (soundId: String) in
      self.clickSoundPlayer.setAccentClickSound(soundId)
    }

    Function("setSubdivisionClickSound") { (soundId: String) in
      self.clickSoundPlayer.setSubdivisionClickSound(soundId)
    }

    Function("previewNormalClick") {
      self.clickSoundPlayer.previewNormalClick()
    }

    Function("previewAccentClick") {
      self.clickSoundPlayer.previewAccentClick()
    }

    Function("previewSubdivisionClick") {
      self.clickSoundPlayer.previewSubdivisionClick()
    }
  }

  private func createMetronomeEngine() -> MetronomeEngine {
    MetronomeEngine(
      clickSoundPlayer: clickSoundPlayer,
      onTick: { [weak self] sequence, beatIndex, beatNumber, beatsPerMeasure, subdivisionIndex, isAccent, timestampMs in
        self?.sendEvent(
          "onTick",
          [
            "sequence": Int(sequence),
            "beatIndex": beatIndex,
            "beatNumber": beatNumber,
            "beatsPerMeasure": beatsPerMeasure,
            "subdivisionIndex": subdivisionIndex,
            "isAccent": isAccent,
            "timestamp": timestampMs,
          ]
        )
      }
    )
  }

  private func readTicksPerBeat(_ value: Any?) -> Int {
    guard let subdivision = value as? String else {
      return 1
    }

    switch subdivision {
    case "eighth":
      return 2
    case "triplet":
      return 3
    case "sixteenth":
      return 4
    default:
      return 1
    }
  }

  private func readTimelineEvents(_ value: Any?) -> [TimelinePlaybackEvent] {
    guard let rawEvents = value as? [Any] else {
      return []
    }

    return rawEvents.compactMap { entry in
      guard let map = entry as? [String: Any] else {
        return nil
      }

      let sequence = Self.readUInt64(map["sequence"]) ?? 0
      let bpm = max(1, Self.readDouble(map["bpm"]) ?? 120)
      let accent = map["accent"] as? Bool ?? false
      let subdivisionIndex = Self.readInt(map["subdivisionIndex"]) ?? 0
      let beatIndexInBar = Self.readInt(map["beatIndexInBar"]) ?? 0
      let beatsPerMeasure = max(1, Self.readInt(map["beatsPerMeasure"]) ?? 4)
      let barId = map["barId"] as? String ?? ""
      let sectionId = map["sectionId"] as? String ?? ""

      return TimelinePlaybackEvent(
        sequence: sequence,
        bpm: bpm,
        accent: accent,
        subdivisionIndex: subdivisionIndex,
        beatIndexInBar: beatIndexInBar,
        beatsPerMeasure: beatsPerMeasure,
        barId: barId,
        sectionId: sectionId
      )
    }
  }

  private static func readDouble(_ value: Any?) -> Double? {
    if let number = value as? NSNumber {
      return number.doubleValue
    }
    return value as? Double
  }

  private static func readInt(_ value: Any?) -> Int? {
    if let number = value as? NSNumber {
      return number.intValue
    }
    return value as? Int
  }

  private static func readUInt64(_ value: Any?) -> UInt64? {
    if let number = value as? NSNumber {
      return number.uint64Value
    }
    if let intValue = value as? Int {
      return UInt64(intValue)
    }
    return value as? UInt64
  }

  private func readAccentPattern(_ value: Any?, beatsPerMeasure: Int) -> [Bool] {
    let pattern: [Bool]? = {
      if let bools = value as? [Bool], !bools.isEmpty {
        return bools
      }

      if let numbers = value as? [NSNumber], !numbers.isEmpty {
        return numbers.map { $0.boolValue }
      }

      return nil
    }()

    guard let pattern else {
      return (0..<beatsPerMeasure).map { index in index == 0 }
    }

    return (0..<beatsPerMeasure).map { index in
      pattern[index % pattern.count]
    }
  }

  private func readSubdivisionAccentMode(_ value: String) -> SubdivisionAccentMode {
    switch value {
    case "group_start":
      return .groupStart
    case "every_nth":
      return .everyNth
    case "custom":
      return .custom
    default:
      return .off
    }
  }
}

File: modules/native-audio/android/src/main/java/expo/modules/nativeaudio/NativeAudioModule.kt
package expo.modules.nativeaudio

import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class NativeAudioModule : Module() {

  private val metronomeEngine: MetronomeEngine by lazy { createMetronomeEngine() }

  private val clickSoundPlayer: ClickSoundPlayer? by lazy {
    appContext.reactContext?.let { ClickSoundPlayer(it) }
  }

  @Volatile
  private var audioInitialized = false

  private val initLock = Any()

  override fun definition() = ModuleDefinition {
    Name("NativeAudioModule")

    Events("onTick")

    OnDestroy {
      clickSoundPlayer?.release()
      audioInitialized = false
    }

    Function("initialize") {
      synchronized(initLock) {
        if (audioInitialized) {
          return@synchronized
        }

        AudioObservability.logBackend()
        clickSoundPlayer?.initialize()
        audioInitialized = true
      }
    }

    Function("areSoundsReady") {
      clickSoundPlayer?.areReady() ?: true
    }

    Function("start") { options: Map<String, Any?> ->
      if (metronomeEngine.running) {
        Log.w(TAG, "start() called while already running — stopping previous loop first")
      }

      val bpm = readDouble(options["bpm"]) ?: 120.0
      val beatsPerMeasure = readInt(options["beatsPerMeasure"]) ?: 4
      val accentPattern = readAccentPattern(options["accentPattern"], beatsPerMeasure)
      val ticksPerBeat = readTicksPerBeat(options["subdivision"])
      val playbackMode = readPlaybackMode(options["playbackMode"])
      val timelineEvents = if (playbackMode == PlaybackMode.SONG_TIMELINE) {
        readTimelineEvents(options["timelineEvents"])
      } else {
        emptyList()
      }

      metronomeEngine.start(
        bpm,
        beatsPerMeasure,
        accentPattern,
        ticksPerBeat,
        playbackMode,
        timelineEvents,
      )
    }

    Function("stop") {
      metronomeEngine.stop()
    }

    Function("setTempo") { bpm: Double ->
      metronomeEngine.updateTempo(bpm)
    }

    Function("setAccentPattern") { accentPattern: List<Boolean> ->
      metronomeEngine.updateAccentPattern(readAccentPatternList(accentPattern))
    }

    Function("setSubdivision") { subdivision: String ->
      metronomeEngine.updateSubdivision(readTicksPerBeat(subdivision))
    }
  }

  private fun createMetronomeEngine(): MetronomeEngine {
    return MetronomeEngine(
      clickSoundPlayer,
      onTick = { sequence,
                 beatIndex,
                 beatNumber,
                 beatsPerMeasure,
                 subdivisionIndex,
                 isAccent,
                 timestampMs ->

        sendEvent(
          "onTick",
          mapOf(
            "sequence" to sequence,
            "beatIndex" to beatIndex,
            "beatNumber" to beatNumber,
            "beatsPerMeasure" to beatsPerMeasure,
            "subdivisionIndex" to subdivisionIndex,
            "isAccent" to isAccent,
            "timestamp" to timestampMs.toDouble(),
          ),
        )
      },
    )
  }

  private fun readPlaybackMode(value: Any?): PlaybackMode {
    return when (value as? String) {
      "song_timeline" -> PlaybackMode.SONG_TIMELINE
      else -> PlaybackMode.QUICK_METRONOME
    }
  }

  private fun readTimelineEvents(value: Any?): List<TimelinePlaybackEvent> {
    val rawEvents = value as? List<*> ?: return emptyList()

    return rawEvents.mapNotNull { entry ->
      val map = entry as? Map<*, *> ?: return@mapNotNull null

      TimelinePlaybackEvent(
        sequence = readLong(map["sequence"]) ?: 0L,
        bpm = readDouble(map["bpm"])?.coerceAtLeast(1.0) ?: 120.0,
        accent = map["accent"] as? Boolean ?: false,
        subdivisionIndex = readInt(map["subdivisionIndex"]) ?: 0,
        beatIndexInBar = readInt(map["beatIndexInBar"]) ?: 0,
        beatsPerMeasure = readInt(map["beatsPerMeasure"])?.coerceAtLeast(1) ?: 4,
        barId = map["barId"] as? String ?: "",
        sectionId = map["sectionId"] as? String ?: "",
      )
    }
  }

  private fun readLong(value: Any?): Long? {
    return when (value) {
      is Int -> value.toLong()
      is Double -> value.toLong()
      is Long -> value
      else -> null
    }
  }

  private fun readDouble(value: Any?): Double? {
    return when (value) {
      is Double -> value
      is Int -> value.toDouble()
      is Float -> value.toDouble()
      else -> null
    }
  }

  private fun readInt(value: Any?): Int? {
    return when (value) {
      is Int -> value
      is Double -> value.toInt()
      else -> null
    }
  }

  private fun readTicksPerBeat(value: Any?): Int {
    return when (value) {
      "eighth" -> 2
      "triplet" -> 3
      "sixteenth" -> 4
      else -> 1
    }
  }

  private fun readAccentPatternList(pattern: List<Boolean>): BooleanArray {
    if (pattern.isEmpty()) {
      return booleanArrayOf(true)
    }
    return pattern.toBooleanArray()
  }

  private fun readAccentPattern(value: Any?, beatsPerMeasure: Int): BooleanArray {
    val pattern = (value as? List<*>)?.map {
      when (it) {
        is Boolean -> it
        else -> false
      }
    }

    if (pattern.isNullOrEmpty()) {
      return BooleanArray(beatsPerMeasure) { index -> index == 0 }
    }

    return BooleanArray(beatsPerMeasure) { index ->
      pattern.getOrElse(index) { index == 0 }
    }
  }

  companion object {
    private const val TAG = "NativeAudioModule"
  }
}
